File: array/merge_intervals/merge_intervals.go
package main

import (
	"fmt"
	"sort"
)

type interval struct {
	start int
	end   int
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
func mergeIntervals(arr []interval) []interval {
	sort.Slice(arr, func(i, j int) bool {
		return arr[i].end < arr[j].end
	})
	n := []interval{arr[0]}
	for _, i := range arr[1:] {
		last := n[len(n)-1]
		if i.start > last.end {
			n = append(n, i)
		} else {
			n[len(n)-1] = interval{last.start, max(i.end, last.end)}
		}
	}
	return n
}

func main() {
	given := [][]int{{1, 3}, {2, 6}, {8, 10}, {15, 18}}
	var givenIntervals []interval
	for _, i := range given {
		givenIntervals = append(givenIntervals, interval{i[0], i[1]})
	}
	expected := [][]int{{1, 6}, {8, 10}, {15, 18}}
	fmt.Println("input: ", given)
	fmt.Println("result: ", mergeIntervals(givenIntervals))
	fmt.Println("output should be: ", expected)
}

File: array/merge_intervals/merge_intervals.rb
def merge_intervals(arr)
  arr.sort! { |a, b| a[0] - b[0] }
  n = [arr[0]]
  arr[1..-1].each do |i|
    last = n[-1]
    if i[0] > last[1]
      n.push(i)
    else
      n[-1] = [last[0], [i[1], last[1]].max]
    end
  end
  n
end

given = [[1, 3], [2, 6], [8, 10], [15, 18]]
expected = [[1, 6], [8, 10], [15, 18]]
p("input: #{given}")
p("result: #{merge_intervals(given)}")
p("output should be: #{expected}")
